micro-rdk-nmea/src/messages/pgns.rs

```diff
@@ -106,3 +115,96 @@ pub struct GnssSatsInView {
     #[length_field = "sats_in_view"]
     satellites: Vec<Satellite>,
 }
+
+macro_rules! define_pgns {
+    ( $(($pgndef:ident, $enum:ident, $pgn:expr)),* ) => {
+        #[derive(Clone, Debug)]
+        pub enum MessageData {
+            $($enum($pgndef)),*,
+            Unsupported(UnparsedMessageData)
+        }
+
+        impl MessageData {
+            pub fn pgn(&self) -> u32 {
+                match self {
+                    $(Self::$enum(_) => $pgn),*,
+                    Self::Unsupported(unparsed) => unparsed.pgn()
+                }
+            }
+
+            pub fn from_bytes(pgn: u32, bytes: Vec<u8>) -> Result<Self, crate::parse_helpers::errors::NmeaParseError> {
+                Ok(match pgn {
+                    $($pgn => {
+                        let cursor = DataCursor::new(bytes);
+                        Self::$enum($pgndef::from_cursor(cursor)?)
+                    }),*,
+                    x => Self::Unsupported(UnparsedMessageData::from_bytes(bytes, x)?)
+                })
+            }
+
+            pub fn to_readings(self) -> Result<GenericReadingsResult, crate::parse_helpers::errors::NmeaParseError> {
+                match self {
+                    $(Self::$enum(msg) => msg.to_readings()),*,
+                    Self::Unsupported(msg) => msg.to_readings()
```

Comment:
Is it meaningful to return an unsupported as a reading?

Reply:
Yeah I think so. If there's a PGN implementation missing this will give you data to test with immediately